### PR TITLE
Improvements to job type

### PR DESCRIPTION
**Pull request by damoxc**
This implements the method as mentioned in the comments previously in the type to return all the jobs in one go.

It also adds support for printing a checksum of the config change, and a diff.

### CODE REVIEW

**Comment by jhoblitt** [comment]
@damoxc It's fantastic to see someone working on the new types!  Are you willing to take a stab at rspec tests?

**Comment by damoxc** [comment]
@jhoblitt Gladly! If I don't get time to look over the weekend then I'll have a look on Monday. Would you like the git history rewritten to include less of my experimentation? Could probably squash it down into a single commit realistically.

**Comment by jhoblitt** [comment]
@damoxc I'm pretty aggressive about squashing in my own workflow, so that a commit is more or less a single feature or bugfix.  Doing so makes it a bit easier to track down the rational for a change with git blame.  That said, it is certainly not a requirement for getting a PR into this module, unless there was something like revert commits or tons of "fix typo" commits (in such cases, the maintainer will often fixup the PR anyways).

TL;DR -- appreciated by not mandated

**Comment by damoxc** [comment]
@jhoblitt  I'll add the rspec tests and rebase into a single commit, or perhaps two (one for the helper, one for checksums/diffs). There's far too much mess I in this I wouldn't want to pollute the mainline repo with.

**Comment by jhoblitt** [comment]
@damoxc I've been using git's `rebase.autosquash=true` feature since around the start of the calendar year, after a particularly ugly squash/re-split session.  This lets me throw in a `fixup! <original title>` or `squash! <original title>` while I'm working without having to squash changes I'm unsure of but also makes it easy to squash down the branch in something suitable of merge.  It's a bit of extra work but I think the reward is worth it -- might be worth a try if you haven't experimented with it before.

**Comment by damoxc** [comment]
@jhoblitt I haven't tried `autosquash` yet, I've always just used `rebase -i`, although I'll look into `autosquash` now I'm aware of it, cheers.

**Comment by jhoblitt** [comment]
@damoxc ping

**Comment by damoxc** [comment]
@jhoblitt Sorry, had a bit of a hectic week last week (job + travels). I started hacking on the rspec tests but going to have to do some more learning around how to test the config property properly, hopefully have some time at the start of this week if all goes well.

**Comment by jhoblitt** [comment]
@damoxc ping I would really like to see this go in.  I actually got a few lines into implementing a `show_diff` before I remember it was already in this PR.

**Comment by damoxc** [comment]
@jhoblitt I had a look yesterday and I'm actually struggling to get the spec tests to run, are there any expected conditions and the environment for them to function?  Current error is

```
Could not autoload puppet/type/user: Could not autoload puppet/provider/user/openbsd: loginclass is not a valid attribute for user on node ***
# ./spec/classes/jenkins_slave_spec.rb:83:in `block (3 levels) in <top (required)>'
```

**Comment by jhoblitt** [comment]
@damoxc That is an odd error.  Are you running the tests on openbsd?

**Comment by damoxc** [comment]
@jhoblitt No, it's on Arch. It goes away if I use Puppet 4.X but then that brings its own slew of issues :smile:

**Comment by jhoblitt** [comment]
@damoxc Are all of your changes pushed?  I've just run into problem (#541) that this should essentially fix.  I am planning to try to deal with this tomorrow morning.

**Comment by damoxc** [comment]
@jhoblitt all my changes are pushed at the moment, that I can see at least.

**Comment by jhoblitt** [comment]
@damoxc I have a working branch but I need to clean it up before merging.

**Comment by jhoblitt** [comment]
@damoxc I have opened PR #544, which consists of this PR + fixes.  I squashed your history and my fixes/refactoring down to a single commit each, so it should be easy to see what changes were required to get this working.

My best guess at this point as to the source of the odd error message you saw, is that it were caused by a puppet methods invoked by your changes that needed to be stubbed out for unit testing.

Thanks for all your hard work!
